File: 10.正则表达式匹配.swift
/*
 * @lc app=leetcode.cn id=10 lang=swift
 *
 * [10] 正则表达式匹配
 */

// @lc code=start
class Solution {
    func isMatch(_ s: String, _ p: String) -> Bool {
        let s = Array(s)
        let p = Array(p)
        let mSize = s.count, nSize = p.count

        let compare = { (idx: Int, jdx: Int) -> Bool in
            if idx == 0 {
                return false
            }
            if p[jdx-1] == "." {
                return true
            }
            return s[idx - 1] == p[jdx - 1]
        }

        var dpTb = [[Bool]](repeating: [Bool](repeating: false, count: nSize + 1), count: mSize + 1)
        dpTb[0][0] = true
        for idx in 0...mSize {
            for jdx in 1...nSize {
                if p[jdx - 1] == "*" {
                    dpTb[idx][jdx] = dpTb[idx][jdx] || dpTb[idx][jdx - 2]
                    if compare(idx, jdx - 1) {
                        dpTb[idx][jdx] = dpTb[idx][jdx] || dpTb[idx - 1][jdx]
                    } 
                } else {
                    if compare(idx, jdx) {
                        dpTb[idx][jdx] = dpTb[idx][jdx] || dpTb[idx - 1][jdx - 1]
                    }
                }
            }
        }
        return dpTb[mSize][nSize]
    }
}
// @lc code=end

File: 10.正则表达式匹配.cpp
/*
 * @lc app=leetcode.cn id=10 lang=cpp
 *
 * [10] 正则表达式匹配
 */

// @lc code=start
class Solution {
public:
    bool isMatch(string s, string p) {
        int m = s.size(), n = p.size();

        auto match = [&](int i, int j) {
            if (i == 0) return  false;
            if (p[j-1] == '.') return true;
            return s[i-1] == p[j-1];
        };

        vector<vector<int>> dp(m+1, vector<int>(n+1));
        dp[0][0] = true;
        for (int i = 0; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (p[j-1] == '*') {
                    dp[i][j] |= dp[i][j-2];
                    if (match(i, j-1)) {
                        dp[i][j] |= dp[i-1][j];
                    }
                } else {
                    if (match(i, j)) {
                        dp[i][j] |= dp[i-1][j-1];
                    }
                }
            }
        }
        return dp[m][n];
    }
};
// @lc code=end

